src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,294 @@
+                    String.format("Improperly escaped '%s' in replacement at col %d: %s", c, x, replacement));
+        }
+    }
+
+    private static int writeReplacementToken(
+            final String replacement,
+            final int offset,
+            final StringBuilder output,
+            final Matcher matcher,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final StringBuilder tokenBuilder,
+            final Map<String, ImmutableList.Builder<String>> variablesUsed) {
+        boolean inReplaceBrackets = false;
+        boolean tokenNumeric = true;
+        tokenBuilder.setLength(0);  // reset the shared builder
+        int x = offset + 1;
+        char c = replacement.charAt(x);
+
+        // Optionally consume the opening brace
+        if (c == '{') {
+            inReplaceBrackets = true;
+            x++;
+            c = replacement.charAt(x);
+        }
+
+        if (inReplaceBrackets) {
+            // Consume until we hit the }
+            while (x < replacement.length() - 1 && c != '}') {
+                if (c == '\\') {
+                    x++;
+                    processEscapedCharacter(replacement, x, tokenBuilder);
+                } else {
+                    tokenBuilder.append(c);
+                }
+                if (tokenNumeric && !Character.isDigit(c)) {
+                    tokenNumeric = false;
+                }
+                x++;
+                c = replacement.charAt(x);
+            }
+            if (c != '}') {
+                throw new IllegalArgumentException("Invalid replacement token, expected '}' at col " + x + ": " + replacement);
+            }
+            x++; // Consume the }
+            output.append(getReplacement(matcher, tokenBuilder.toString(), variables, variablesUsed));
+        } else {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Invalid replacement token, tokens must be surrounded by { } at col %d: %s",
+                                x,
+                                replacement));
+        }
+        return x - offset - 1;
+    }
+
+    private static String getReplacement(
+            final Matcher matcher,
+            final String replaceToken,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final Map<String, ImmutableList.Builder<String>> variablesUsed) {
+        final int prefixSeparatorIndex = replaceToken.indexOf(':');
+        final String prefix;
+        final String variableName;
+        if (prefixSeparatorIndex == replaceToken.length() - 1) {
+            throw new IllegalArgumentException(
+                    String.format("found prefix in variable replacement, but no variable name found: '%s'",
+                    replaceToken));
+        }
+        if (prefixSeparatorIndex != -1) {
+            prefix = replaceToken.substring(0, prefixSeparatorIndex);
```

Comment:
What happens if I have `${:foo}`; won't that be processed as non-prefixed? Where is prefix validation done?

Reply:
(empty prefix is not the same as no prefix)

Reply:
The way these are processed, empty prefix is the same as no prefix.